File: frontend/user/lib/domain/usecase/product/get_products_usecase.dart
import 'package:b2205946_duonghuuluan_luanvan/core/error/failure.dart';
import 'package:dartz/dartz.dart';
import 'package:b2205946_duonghuuluan_luanvan/domain/entity/product/product.dart';
import 'package:b2205946_duonghuuluan_luanvan/domain/repository/product/product_repository.dart';

class GetProductsUseCase {
  final ProductRepository _repo;
  GetProductsUseCase(this._repo);
  Future<Either<Failure, List<Product>>> call({int? categoryId, int? page, int? perPage, String? keyword}) =>
      _repo.getAllProduct(categoryId: categoryId, page: page, perPage: perPage, keyword: keyword);
}

File: backend/app/application/use_case/product/get_products_usecase.py
from typing import Optional

from app.domain.repositories.product_repository import ProductRepository


class GetProductsUseCase:
    def __init__(self, repo: ProductRepository):
        self.repo = repo

    def execute(self, page: int = 1, per_page: Optional[int] = None,
                keyword: Optional[str] = None, category_id: Optional[int] = None) -> dict:
        return self.repo.get_all_with_details(
            page=page, per_page=per_page,
            keyword=keyword, category_id=category_id,
        )
